### PR TITLE
HTCONDOR-727 If an EP starts in a cgroup with a memory limit, advertise

### PR DESCRIPTION
**Pull request by GregThain**
that amount of memory for jobs to use.

<Insert PR description here, and leave checklist below for code review.>

# HTCondor Pull Request Checklist for internal reviewers

- [ ] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [ ] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [ ] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [ ] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [ ] Verify that the branch destination of the PR matches the target version of the ticket
- [ ] Check for correctness of change
- [ ] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [ ] Check for documentation, if needed  (documentation [build logs](https://readthedocs.org/projects/htcondor/builds/))
- [ ] Check for version history, if needed
- [ ] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [ ] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

File: src/condor_sysapi/phys_mem.cpp [hunk 1]
```diff
@@ -24,21 +24,91 @@
...
+#include <cstdio>
+#include <algorithm>
+
+
+// Assuming a cgroup-style file with an integer
+// or a string that we don't care about in it,
+// return the 64 bit int or 0 if it is a string.
+int64_t file_size_contents(const char *filename) {
+	int64_t s = 0;
+	FILE *f = fopen(filename, "r");
```
Comment (JaimeFrey) [inline src/condor_sysapi/phys_mem.cpp:36]:
```suggestion
	FILE *f = safe_fopen_wrapper_follow(filename, "r");
```

File: src/condor_sysapi/phys_mem.cpp [hunk 2]
```diff
@@ -24,21 +24,91 @@
...
+	if (s > (1ll << 60)) {
+		s = 0;
+	}
+	return s;
+}
+
+int64_t cgroup_current_memory_limit() {
+	// Note all these files should be readable
+	// by self without boosting privileges.
+	FILE *f = fopen("/proc/self/cgroup", "r");
```
Comment (JaimeFrey) [inline src/condor_sysapi/phys_mem.cpp:55]:
```suggestion
	FILE *f = safe_fopen_wrapper_follow("/proc/self/cgroup", "r");
```

File: src/condor_sysapi/phys_mem.cpp [hunk 3]
```diff
@@ -24,21 +24,91 @@
...
+// or a string that we don't care about in it,
+// return the 64 bit int or 0 if it is a string.
+int64_t file_size_contents(const char *filename) {
+	int64_t s = 0;
+	FILE *f = fopen(filename, "r");
+
+	if (f != nullptr) {
+		std::ignore = fscanf(f, "%zd", &s);
+		fclose(f);
+		return s;
```
Comment (JaimeFrey) [inline src/condor_sysapi/phys_mem.cpp:41]: Is this return meant to be here? It makes the conditional below into dead code.

**Review by JaimeFrey — CHANGES_REQUESTED** [review]

File: src/condor_sysapi/phys_mem.cpp (hunk 1 above)
Comment (GregThain) [reply src/condor_sysapi/phys_mem.cpp:36]: Not sure we need this, given that /cgroup filesystem is guaranteed not to have internal symlinks or other nonsense that safe_... protects against.

Comment (JaimeFrey) [reply src/condor_sysapi/phys_mem.cpp]: If the filename is full path that's a constant in the same function, sure. When you're passing the filename in as a parameter, it's too easy for another caller to creep in unnoticed with a less-safe value.

**Review by JaimeFrey — APPROVED** [review]
LGTM

**Pull request merged**